File: algorithms/682_baseball_game.py
'''
  LeetCode.com
  Problem Number: 682
  Problem: baseball game
  Rakesh Kumar @ cpp.rakesh@gmail.com
  Date: 18/05/2020
'''

from typing import List

class Solution:
    def calPoints(self, ops: List[str]) -> int:
        d = [0] * len(ops)
        i = 0

        for o in ops:
            if o == '+':
                d[i] = d[i - 1] + d[i - 2]
                i += 1
            elif o == 'D':
                d[i] = d[i - 1] << 1
                i += 1
            elif o == 'C':
                d[i - 1] = 0
                i -= 1
            else:
                d[i] = int(o)
                i += 1

        s = 0
        for score in d:
            s += score
        return s


if __name__ == '__main__':
    s = Solution()
    print(s.calPoints(["5", "2", "C", "D", "+"]))
    print(s.calPoints(["5", "-2", "4", "C", "D", "9", "+", "+"]))

File: 682_baseball_game.cpp
/*
  LeetCode.com
  Problem Number: 682
  Problem: baseball game
  Rakesh Kumar @ cpp.rakesh@gmail.com
  Date: 18/05/2020
*/

#include <bits/stdc++.h>

class Solution {
public:
    int calPoints(std::vector<std::string>& ops) {
        std::stack<int> d;
        for (const std::string& s : ops) {
            if (s == "+") {
                const int t = d.top();
                d.pop();
                const int n = d.top() + t;
                d.push(t);
                d.push(n);
            } else if (s == "D") {
                d.push(d.top() * 2);
            } else if (s == "C") {
                d.pop();
            } else {
                d.push(std::atoi(s.c_str()));
            }
        }

        int r = 0;
        while (d.empty() == false) {
            r += d.top();
            d.pop();
        }

        return r;
    }
};

int main() {
    Solution s;
    std::vector<std::string> v = {"5", "2", "C", "D", "+"};
    printf("%d\n", s.calPoints(v));

    v = {"5", "-2", "4", "C", "D", "9", "+", "+"};
    printf("%d\n", s.calPoints(v));

    return 0;
}
